hugr-passes/src/dataflow/datalog.rs

```diff
@@ -0,0 +1,348 @@
+//! [ascent] datalog implementation of analysis.
+
+use ascent::lattice::BoundedLattice;
+use itertools::Itertools;
+
+use hugr_core::extension::prelude::{MakeTuple, UnpackTuple};
+use hugr_core::ops::{OpTrait, OpType};
+use hugr_core::{HugrView, IncomingPort, Node, OutgoingPort, PortIndex as _, Wire};
+
+use super::value_row::ValueRow;
+use super::{AbstractValue, AnalysisResults, DFContext, PartialValue};
+
+type PV<V> = PartialValue<V>;
+
+/// Basic structure for performing an analysis. Usage:
+/// 1. Get a new instance via [Self::default()]
+/// 2. (Optionally / for tests) zero or more [Self::prepopulate_wire] with initial values
+/// 3. Call [Self::run] to produce [AnalysisResults]
+pub struct Machine<V: AbstractValue>(Vec<(Node, IncomingPort, PartialValue<V>)>);
+
+/// derived-Default requires the context to be Defaultable, which is unnecessary
+impl<V: AbstractValue> Default for Machine<V> {
+    fn default() -> Self {
+        Self(Default::default())
+    }
+}
+
+impl<V: AbstractValue> Machine<V> {
+    /// Provide initial values for some wires.
+    // Likely for test purposes only - should we make non-pub or #[cfg(test)] ?
+    pub fn prepopulate_wire(&mut self, h: &impl HugrView, wire: Wire, value: PartialValue<V>) {
+        self.0.extend(
+            h.linked_inputs(wire.node(), wire.source())
+                .map(|(n, inp)| (n, inp, value.clone())),
+        );
+    }
+
+    /// Run the analysis (iterate until a lattice fixpoint is reached),
+    /// given initial values for some of the root node inputs.
+    /// (Note that `in_values` will not be useful for `Case` or `DFB`-rooted Hugrs,
+    /// but should handle other containers.)
+    /// The context passed in allows interpretation of leaf operations.
+    pub fn run<C: DFContext<V>>(
+        mut self,
+        context: C,
+        in_values: impl IntoIterator<Item = (IncomingPort, PartialValue<V>)>,
+    ) -> AnalysisResults<V, C> {
+        let root = context.root();
+        self.0
+            .extend(in_values.into_iter().map(|(p, v)| (root, p, v)));
+        // Any inputs we don't have values for, we must assume `Top` to ensure safety of analysis
+        // (Consider: for a conditional that selects *either* the unknown input *or* value V,
+        // analysis must produce Top == we-know-nothing, not `V` !)
+        let mut have_inputs =
+            vec![false; context.signature(root).unwrap_or_default().input_count()];
+        self.0.iter().for_each(|(n, p, _)| {
+            if n == &root {
+                if let Some(e) = have_inputs.get_mut(p.index()) {
+                    *e = true;
+                }
+            }
+        });
+        for (i, b) in have_inputs.into_iter().enumerate() {
+            if !b {
+                self.0
+                    .push((root, IncomingPort::from(i), PartialValue::Top));
+            }
+        }
+        // Note/TODO, if analysis is running on a subregion then we should do similar
```

Comment:
I'll raise an issue for this. It's not clear how to identify nonlocal edges into a subregion, or to differentiate between nonlocal edges whose source we can't see, and ports with no edge at all, where perhaps "Bottom" is appropriate (it really can't happen, that bit of Hugr can't run....).